[DM-29923] Initial version

rra opened the pull request.

rra committed:
Starting point with introduction and threat model

rra committed:
Add authentication and isolation sections

rra committed:
Added CSRF and XSS sections

rra committed:
Add implementation status and some more notes

frossie commented on index.rst:
I know what you mean and that  it is the correct use of the term, but you might want to add "personal" or some other modifier to "confidential data" to make sure the reader doesn't just think of prorprietary user data here. 
(I know this is also addressed at length before but people who skim read are drawn to bullet points)

frossie commented on index.rst:
Not a real comment, but as an aside: allowed scoped may also very by time, for example if we are flooded with users and have limited resources we may say that "for the next 24 hours, only users of group X can have access to nublado". Just something to think about when we are thinking about (a) token renewal/expiration and (b) how we present authorsiation fail messages to the user

rra replied:
Good point, I haven't been thinking about rate limiting that way but it's a good way to think about it. I don't think it fits into this document, but I'll keep it in mind.

frossie commented on index.rst:
minor: the notebook-portal reference does not resolve

frossie commented on index.rst:
I don't know if it's worth explicitly discussing this, but what if we had a saphir/roundtable service for microservices or 3rd party-to-RSP but codeployed apps (like the alert filtering service)? Not sure whether we should demand for all of them to be in separate origins, or an origin for all these "guest" services (and also unclear as to whether this would make the safir/templating layer's job harder or not"

rra replied:
They can all share an origin with trusted APIs as well, as long as nothing on that origin serves any JavaScript. I'll add another paragraph explaining this. The only thing we need to separate into its own origins is services that serve a browser UI.

rra committed:
Additional notes from review feedback

- Fix one cross-reference
- Add a paragraph about API origins
- Clarify what was meant by confidential data

jonathansick commented on index.rst:
I think both of these are missing their links.

jonathansick reviewed (approved):
### Question: mixing authenticated and unauthenticated traffic to a backend?

Does our authentication model allow for both authenticated and *un*authenticated requests to the same backend service (either to the same endpoints, or to a subset of endpoints)? Say I'm building the Semaphore (notification service) API. If a visitor to Squareone isn't logged in, Squareone will still need to get data from Semaphore (global messages, like "Maintainance in X hours" or "Data Release X is now available! Read more..."). However, if the user is logged in, I can imagine that the response might include additional data (targetted messages to a user). So the same endpoint might show more data, or additional endpoints altogether might become available (like a POST endpoint where a user can mark a message as read). In this case, will the custom ingress configuration resource narrow down a detailed list of endpoints accessible to unauthenticated users? Or is this a case where we need the application to exercise its own authorization logic based on the authentication data? For GraphQL APIs, the app would definitely have to do its own authorization.

### Comment: Will CORS be more common than we expect?

The technote talks about Notebook to Portal CORS, but I'm wondering if CORS will become more frequent? Take Semaphore, again. Squareone, Notebook Aspect, and Portal might all want to use the semaphore API to implement a UI feature (I don't know if JupyterLab would be using it via JS or Python, though).

rra committed:
Add missing CORS and fetch links

rra committed:
Add unauthenticated section, flesh out cross-origin

Add a new section on unauthenticated and optionally authenticated
routes that specifes the credential isolation measures needed for
those routes.  Rewrite the section on Notebook to Portal calls to
be more generic and describe any cross-origin requests internal to
the Rubin Science Platform.

rra commented:
These are great comments. Thank you!

I added a new section explicitly addressing unauthenticated routes and optionally authenticated routes. We may need to add some new features to Gafaelfawr to support them, but it should be possible to do. A possible implementation approach is now spelled out in the docs.

Good point that cross-origin requests may not be as rare as I was hoping. I rewrote the section on Notebook to Portal cross-origin requests to be more generic.

ktlim commented on index.rst:
```suggestion
This, in turn, will be used to investigate possible security incidents and to look for anomalies in how tokens are used (such as use from an unexpected IP address).
```

ktlim commented on index.rst:
Some APIs might return user-computed files or user-computed objects from an object store.  I'm thinking WebDAV, but the "remote Butler" service could also do this.  These would presumably need separate origins as well, unless we can ensure that they are always returned in such a way that they are not rendered by a browser.

(To be clear, I don't think any additional text is needed for this; I'm just checking that I understand correctly.)

rra replied:
Good catch, I wasn't thinking of that. I think this is worth explicitly calling out since there's also a `Content-Security-Policy` header that should be added for such APIs. I've added some additional wording for that.

ktlim commented on index.rst:
```suggestion
Unless required by a protocol that the application needs to implement, only applications intended for use via a web browser should accept ``POST`` with a ``Content-Type`` of ``application/x-www-form-urlencoded``, ``multipart/form-data``, or ``text/plain``.
```

ktlim commented on index.rst:
```suggestion
The browser will add the ``Origin`` header automatically to cross-origin (and some same-origin) requests, and it cannot be disabled in JavaScript.
```

ktlim commented on index.rst:
```suggestion
If so, it must respond with success, copying the ``Origin`` value to the ``Access-Control-Allow-Origin`` response header and including ``Access-Control-Allow-Credentials: true`` in the response headers.
```

ktlim commented on index.rst:
```suggestion
When the application is written internally by Rubin Observatory, there is no reason not to have it send its own ``Content-Security-Policy`` header.
```

rra committed:
Fix spelling error

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Fix spelling error

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Fix spelling error

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Fix spelling error

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Fix spelling error

Co-authored-by: Kian-Tat Lim <ktl@slac.stanford.edu>

rra committed:
Add recommendation for APIs serving user data

An API service that returns raw user-controlled data needs its own
origin and a restrictive Content-Security-Policy.
